### ISSUE TITLE
Scaling Rips persistence on 7D point clouds from time series (100–10,000+ points)

### ISSUE THREAD

**Issue by DaliSanaa** [GUDHI/gudhi-devel#1325]
Hi,

I'm working on a classification system that analyzes time series signals by embedding them as point clouds and computing persistent homology. GUDHI handles the topology computation and works well at small scale. Looking for guidance on scaling up.

**Setup**

My signals are repetitive sequences (think repeated cycles of a process). Each cycle produces a segment of ~10-20 data points, each embedded in 7D feature space (3 structural features + 2 rate-of-change features + 2 cumulative trajectory features). All cycles of one signal are combined into a single point cloud, and I run Rips persistence to extract the topological signature (H0, H1, H2). The topology discriminates between different signal types — I've validated 9 distinct discrimination dimensions across 19 test signals.

**Parameters**

RipsComplex(points, max_edge_length=50.0)
create_simplex_tree(max_dimension=3)
4 of 7 dimensions normalized to [-1, 1], remaining 3 are categorical/positional (range 0-18)
Near-complete connectivity at max_edge_length=50
Current scale (works)

10 cycles × 13 points = ~130 points in 7D
Completes in seconds
Where it breaks

100 cycles × 13 points = ~1,300 points in 7D
Does not complete within minutes
Root cause: near-complete connectivity creates a massive simplex tree
Current workaround

I cluster similar cycles and send only cluster medoids to GUDHI, reducing ~1,300 points to ~140. This completes in ~7 seconds but loses topological detail. The full cloud has richer H1/H2 structure than the compressed version.

**Target scale**

Production use requires thousands of cycles, potentially tens of thousands of points. The system needs to learn continuously from new signals.

**Data characteristics that might help**

Points are not randomly distributed — they follow structured trajectories (linear ramps, exponential curves, oscillations)
Cycles are repetitive with gradual variation (early cycles similar, late cycles similar, drift between them)
The discriminating topological features (H1 cycles, H2 voids) tend to have low persistence values — they live at small scales
Large-scale features (high persistence) are mostly H0 component merging, which is the least informative for my application
Constraint

I need exact persistence, or an approximation with a bounded and controllable error. The H1 cycle counts and H2 void detection are the discriminating signal — losing them to aggressive approximation would break classification.

**Questions**

For the 1,300-point case: would edge collapse solve this while preserving exact persistence? My understanding is that it removes dominated edges without changing the persistence diagram.
For 10,000+ points: what GUDHI approach would you recommend — sparse Rips with tight epsilon, a different complex type, or a different strategy entirely?
Given that my data has low-dimensional structure (trajectories in 7D, not uniformly distributed), is there a more efficient filtration than Vietoris-Rips that would exploit this?
Would reducing max_edge_length be safe if the discriminating features live at small scales? I haven't tested this systematically yet.
Happy to share sample point clouds or more detail about the data distribution if that helps.

Thanks for building GUDHI

**Comment by DavidLapous** [comment]
Hi @DaliSanaa,
Thank you for using Gudhi 😉
> would edge collapse solve this while preserving exact persistence?

Yes, the RipsComplex is a flag complex, so this won't change the homology. Furthermore, doing these collapses **before** expanding the complex will be much more efficient, as much less simplices will be built.
I'd suggest something like
```python
st = gd.RipsComplex(points, max_edge_length=50.0).create_simplex_tree(max_dimension=1)
st.collapse_edges(100) # only the first few are "effective" but the more you expand the more it will help
st.expansion(3)
dgms = st.persistence()
```
if you want to do it manually, but we provide pipelines that sometime offer faster direct implementation (for instance, `dgms` can sometime be more efficiently be computed without even building `st`):

```python
from gudhi.sklearn.rips_persistence import RipsPersistence
dgms, = RipsPersistence(threshold=50, num_collapses=13, homology_dimensions=[0,1,2]).fit_transform([points])
```

> Given that my data has low-dimensional structure (trajectories in 7D, not uniformly distributed), is there a more efficient filtration than Vietoris-Rips that would exploit this?

I'm not aware of that. Exploiting the geometry in 7d is hard as Delaunay complexes are hard to compute there. 

> Would reducing max_edge_length be safe if the discriminating features live at small scales?

Yes if "small" < `max_edge_length`. 

> For 10,000+ points: what GUDHI approach would you recommend — sparse Rips with tight epsilon, a different complex type, or a different strategy entirely?

If the code I gave above is not necessary, this might be useful yes. But at 10k points I'm not sure this is necessary.
